apis/Google.Cloud.Diagnostics.AspNetCore/Google.Cloud.Diagnostics.AspNetCore/Trace/CloudTraceExtension.cs

```diff
@@ -163,35 +160,19 @@ internal static TraceHeaderContext CreateTraceHeaderContext(IServiceProvider pro
         {
             var accessor = provider.GetServiceCheckNotNull<IHttpContextAccessor>();
             var shouldTraceRequest = provider.GetServiceCheckNotNull<ShouldTraceRequest>();
-            var traceIdFactory = provider.GetServiceCheckNotNull<TraceIdFactory>();
 
             string header = accessor.HttpContext?.Request?.Headers[TraceHeaderContext.TraceHeader];
-            var traceHeaderContext = TraceHeaderContext.FromHeader(header);
-            if (traceHeaderContext.ShouldTrace == null && shouldTraceRequest.ShouldTrace != null)
-            {
-                bool? shouldTrace = shouldTraceRequest.ShouldTrace(accessor.HttpContext?.Request);
-                if (shouldTrace == true)
-                {
-                    return new TraceHeaderContext(
-                        traceHeaderContext.TraceId ?? traceIdFactory.NextId(),
-                        traceHeaderContext.SpanId ?? 0, shouldTrace);
-                }
-                else if (shouldTrace == false)
-                {
-                    return new TraceHeaderContext(
-                        traceHeaderContext.TraceId, traceHeaderContext.SpanId, shouldTrace);
-                }
-            }
-            return traceHeaderContext;
+            Func<bool?> shouldTraceFunc = () => shouldTraceRequest?.ShouldTrace(accessor.HttpContext?.Request);
+            return TraceHeaderContext.FromHeader(header, shouldTraceFunc);
         }
 
         /// <summary>
-        /// Creates a <see cref="ContextManagedTracer"/>.
+        /// Creates a <see cref="DelegatingTracer"/>.
```

Comment:
Do we actually need to be specific here? Why not just make it return `IManagedTracer` (which means you can remove the type arguments to AddSingleton) and document the purpose of the method rather than the implementation?